Implementation of TreeSearch for TKF models

MattesMrzik opened the pull request:
Closes: #115, including #122. 
Will do #123 in another PR

MattesMrzik committed:
firsttry TreeSearchCost+reestimate,slightly buggy
- update maps doesnt update the seqs only the mappings: bug: they are
  not in sync
- still more tests are needed
- udpate tree is not checked if that impl is actually correct

MattesMrzik committed:
removed old test trait

github-actions[bot] committed:
Format Rust code using rustfmt

MattesMrzik committed:
spelling

MattesMrzik committed:
reestimate with dp and bruteforce+multithreading

MattesMrzik committed:
fixed: inval childs, update correct maps, multi
- internal tmp mode info needs to be set invalid for childs and sibling
  of v2
- from iter for fixed bit set behaves not like my intuition
- multi now does not copy cost for every possibility, instead only justonce
  once per thread

MattesMrzik committed:
fix:makeValid postReestimate forgot reset prevDelOrNot

github-actions[bot] committed:
Format Rust code using rustfmt

MattesMrzik committed:
save brute force maxes in many files,
- can be recomputed with feature
- recomputation can be done with multi threads enabled by feature

MattesMrzik committed:
Merge branch '115-TKF-tree-seach' of github.com:MattesMrzik/rust-phylo into 115-TKF-tree-seach

github-actions[bot] committed:
Format Rust code using rustfmt

MattesMrzik committed:
tkf92:fixed fragmentation matches simulation logl

MattesMrzik committed:
comment to, edge seqs from assign possibilities

MattesMrzik committed:
renamed to brute_force_ancestors.rs

MattesMrzik committed:
brute force max a single save file, and cleanup

MattesMrzik committed:
tkf92_fixed cost builder with checks,
- removed some of the simple reestimate tests, since i have the big with
  the save file anyway

MattesMrzik committed:
renamed to tkf92_fixed_fragmentation

MattesMrzik committed:
some moving around of filenames and test fns

MattesMrzik committed:
more clean up, all tkf pass, with reestimate from file

MattesMrzik committed:
a comment for test

github-actions[bot] committed:
Format Rust code using rustfmt

MattesMrzik committed:
rng to reestimate and some cleanup

MattesMrzik committed:
remove the tree_ancestor_co_estimation trait

MattesMrzik committed:
added some comments

MattesMrzik committed:
fn phylo_info.check_dollos_constraint

MattesMrzik committed:
special tkf models, and nicer comments
- tkf model with fixed fragmentation, used to compare to simulation
  prob, and to check whether the integration works
- tkf model with additional blocks, used by the reestimation tests,
  since reestimation might change the msa to have different blocks
-

MattesMrzik committed:
cleanup in reestimate

MattesMrzik committed:
brute force max working

MattesMrzik committed:
Merge branch '115-TKF-tree-seach' of github.com:MattesMrzik/rust-phylo into 115-TKF-tree-seach

github-actions[bot] committed:
Format Rust code using rustfmt

MattesMrzik committed:
Merge remote-tracking branch 'upstream/develop' into 115-TKF-tree-seach

MattesMrzik committed:
removed debug assert that failed tests

MattesMrzik committed:
ignore some test when coverage is run

MattesMrzik committed:
fix update ancestral map

github-actions[bot] committed:
Format Rust code using rustfmt

MattesMrzik committed:
triggert tests

MattesMrzik committed:
more tests and added #[cfg_attr(coverage, coverage(off))] to tests

github-actions[bot] committed:
Format Rust code using rustfmt

MattesMrzik committed:
more cleanup

MattesMrzik committed:
changed data to be msa with changing fragments

junniest commented on phylo/data/tkf/fixed_fragments/masa.fasta:
```diff
@@ -0,0 +1,15 @@
+>R
+--------------NN------NNNN--------------N--------------NNN----------------N-NNNN-----------NN-----------N------N------N----------NNNNNN-N--NN---NNNNN---NNNNNNNN-----------------------NN--------NNNNNN------N--------------NNN-NN-N------N----N--------NNN---NNN----------------------------N-------------N------NN-N--------------NNNNNNN------------------NNNNN---------NN----NNN----
```
If I understand this correctly, the N here stands for an ambiguous symbol? This will work fine for DNA alignments, however in proteins N stands for Asparagine, so it's a very specific AA. I am not sure which kind of alignment this file is meant to be, so I suggest either changing the char to X if the type is irrelevant, or making it clear in the file name that this is a DNA alignment (or, maybe even better, both).

junniest commented on phylo/Cargo.toml:
```diff
@@ -26,6 +26,8 @@ precomputed-test-results = []
 par-regraft = ["dep:rayon"]
 par-regraft-chunk = ["dep:rayon"]
 par-regraft-manual = ["dep:rayon"]
+recompute-brute-force-max-ancestral-seqs = []
```
this feature name seems excessively long lol. Would maybe `recompute-brute-force-ancestors` be enough? I think it's clear you are maximising the asr... Not sure if this is important enough to bother in the first place, but it feels excessive.

junniest commented on phylo/Cargo.toml:
```diff
@@ -26,6 +26,8 @@ precomputed-test-results = []
 par-regraft-chunk = ["dep:rayon"]
 par-regraft-manual = ["dep:rayon"]
+recompute-brute-force-max-ancestral-seqs = []
+multi-thread = ["dep:rayon"]
```
This is a curious addition. And it probably makes no sense to use multi-threading here but not use either of the par-regraft features. I think that in the future I will remove the `par-regraft-whatever` features entirely, replacing them with this one and keeping the most efficient spr parallelisation impl. Although if I remember correctly there's some trade-offs between different modes, but the fact that there's three parallel code variants + the single thread one is too annoying to me.

junniest replied:
The other question is whether it should be a compile feature in the first place or if the thread number should be a call parameter...

MattesMrzik replied:
This is compile feature is only relevant for the brute force max tests. We could also just default to using rayon and multi threading compilation for the tests and in the case that either the pre-calculated file is not present or `recompute-brute-force-ancestors` is passed then use multiple threads by default to run the brute force max.

junniest replied:
Yeah, I understand. I think that the feature name is somewhat misleading though, since it's only meant for tests, but can seem like it's meant for all computation. I think it should still be optional even for tests, so I don't think we should force it.
Maybe rename it to `multithread-brute-force-asr`? Just to make sure it's clear what this feature is relevant for.

junniest commented on phylo/data/tkf/reestimate/README.txt:
```diff
@@ -0,0 +1,7 @@
+It is used for checking the correctness of the Dynamic Programming (DP) algorithm that re-estimates 
+the ancestral sequences of two adjacent nodes in a phylogenetic tree under the TKF92 indel model.
+The DP is checked against a brute-force calculation of the optimal ancestral sequences. Precomputed 
+results of the brute-force calculation are provided, but can be recomputed by passing '--recompute-brute-force-max-ancestral-seqs'
```
you need to pass `--features recompute-brute-force-max-ancestral-seqs` rather than the feature name directly (but I also suggest shortening the name below). Same for `--features multi-thread`.

MattesMrzik replied:
Thanks, i adjusted the text.

junniest commented on phylo/data/tkf/reestimate/tree.newick:
```diff
@@ -0,0 +1,2 @@
+((((L0:0.81078,(L1:0.55514,(L2:0.4338,(L3:0.18962,L4:0.17389)N192:0.3372)N193:0.09462)N194:0.12018)N195:0.15444,((L9:0.10412,L10:0.09785)N196:0.66151,((L12:0.3466,(L13:0.61597,(L14:0.02602,L15:0.03087)N197:0.30588)N198:0.12801)N199:0.10395,((L19:0.02366,(L20:0.02786,L21:0.02179)N200:0.02239)N201:0.3622,((L24:0.28592,L25:0.13226)N202:0.23828,(L27:0.35331,L28:0.36583)N203:0.07446)N204:0.06284)N205: ...
```
damn this tree is big lol. Did you do brute force ASR for this huge thing?.. That may be an overkill for testing purposes, possibly...

MattesMrzik replied:
Yep, it only takes a couple minutes and only about 6 sec if the precomputed results (included in git) are used. If this is acceptable then I would keep it like it is. If not i can try to extract a subtree from this tree and alignment. Because it has the property that reestimation of the seqs changes the observed fragment boundaries, which is a nice thing to test.

junniest replied:
A few minutes is fine, I think! Cool that it's so fast actually. Do you think it should be a unit test or an integration test? Maybe a smaller unit test would be helpful and this could move to integration.

MattesMrzik replied:
Implemented more unit tests.

junniest commented on phylo/src/alignment/mod.rs:
```diff
@@ -351,6 +352,10 @@ pub struct MASA {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         let both_maps = self.leaf_maps.iter().chain(self.ancestral_maps.iter());
+        // sort by node id
+        let both_maps = both_maps.sorted_by(|(node_idx_a, _), (node_idx_b, _)| {
```
Suggestion: using `sorted_by_key` would be cleaner here, i.e.:
```
let both_maps = both_maps.sorted_by_key(|(node_idx, _)| {
    &self.idx_to_id[usize::from(**node_idx)]
});
```

MattesMrzik replied:
see #135

junniest commented on phylo/src/alignment/mod.rs:
```diff
@@ -351,6 +352,10 @@ pub struct MASA {
 impl Display for MASA {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         let both_maps = self.leaf_maps.iter().chain(self.ancestral_maps.iter());
```
One question here is whether we would like to keep the original order of sequences (or at least have the leaves printed first, then internal nodes, then root?). If, for example, the node IDs are `IA` and `IB` instead of `A` and `B`, the output becomes this (with internal sequences first):
```
>I01
A-CCA
>I02
---AT
>IA
--A--
>IB
-CA--
>IC
--X-G
>ID
T-CAA
>Root
TATTG
```
The root is only last here because `ID<Root`.

I am not sure what is best here. Trying to maintain the order which we read the sequences in seems like to much work. But potentially first printing the leaves, then the internal nodes and the root last could be reasonable?

MattesMrzik replied:
please see #135

junniest commented on phylo/src/alignment/mod.rs:
```diff
@@ -469,11 +474,34 @@ impl AncestralAlignment for MASA {
         if let Some(anc_map) = self.ancestral_maps.get_mut(node_idx) {
+            let id = &self.idx_to_id[usize::from(*node_idx)];
+            let old_record = self.ancestral_seqs.record_by_id(id);
+            let old_seq = old_record.seq();
```
Nitpick: these two lines could be merged to remove the `old_record` var.

MattesMrzik replied:
Down below i am again using `old_record` to get its description.

junniest replied:
I see, sorry!

junniest commented on phylo/src/alignment/mod.rs:
```diff
@@ -469,11 +474,34 @@ impl AncestralAlignment for MASA {
+            let id = &self.idx_to_id[usize::from(*node_idx)];
+            let old_record = self.ancestral_seqs.record_by_id(id);
+            let old_seq = old_record.seq();
+            let mut new_seq = Vec::new();
```
Since you know the exact length of the new record (shouldn't it be the same as `old_seq.len()`?) you should use `Vec::with_capacity(len)` to avoid memory allocation as the vector is expanded.

MattesMrzik replied:
I know how long the new mapping is, but not how long the new seq. the new mapping is for example `None, Some(0), None, Some(1), None, None`, then the new seq leg is equal to the number of Somes or alternatively the usize in the last Some.

junniest replied:
I see. I would then suggest something like this:
`let mut new_seq = Vec::with_capacity(new_map.len());`
and then calling
`new_seq.shrink_to_fit();` at the end.

`Vec::new()` doesn't allocate memory at all, meaning that when you push it will check for capacity and reallocate memory. This will happen relatively efficiently, it's not doing this every time, but it will happen multiple times anyway. If you allocate with an upper bound and then shrink to fit, it will be at most two memory allocations (once at the start and potentially at the shrinking point if it's better to store the Vec elsewhere).

It's a minor thing, but optimising memory allocations is quite fun (for me at least lol), and can sometimes drastically affect performance. Probably not here, but it makes sense to think about it anyway, I think.

junniest commented on phylo/src/alignment/mod.rs:
```diff
@@ -469,11 +474,34 @@ impl AncestralAlignment for MASA {
+                } else if new_site.is_some() {
+                    new_seq.push(AMB_CHAR);
+                }
+            }
```
This whole block is very confusing to read, I am not sure I completely understand what's going on :D 
I've been thinking of switching to bitsets for storing maps for a while, would using that make this easier? The only difference would be that this implementation (with a `Vec<Option<usize>>`) lets us switch sites around, so something like `Some(2), Some(1), Some(0)` could represent an inversion. We never use this and I wonder if it's actually a bad idea to have this option in the first place.

MattesMrzik replied:
I added some comment to make clear what this code does.

junniest commented on phylo/src/alignment/mod.rs:
```diff
@@ -469,11 +474,34 @@ impl AncestralAlignment for MASA {
+                }
+            }
             *anc_map = map;
+            let new_record = record!(id, old_record.desc(), &new_seq);
```
This definitely allocates a new record object. If the id and description should stay the same, wouldn't it make more sense to just replace the sequence?

MattesMrzik replied:
I think records are not modifiable: https://docs.rs/bio/latest/bio/io/fasta/struct.Record.html

junniest replied:
Ah damn. That sucks, I see. I did not think of that. We might want to add an internal representation eventually then (just of the seq array), just so that we don't have to allocate new records every time. 
Actually, would internal sequences ever have a description? I guess they might if they are provided in the original MASA.

MattesMrzik replied:
See #143

junniest commented on phylo/src/alignment/sequences.rs:
````diff
@@ -176,6 +176,33 @@ impl Sequences {
+    /// seqs.update_record("seq1", new_record);
+    /// assert_eq!(seqs.record_by_id("seq1").seq(), b"C");
+    /// ```
+    pub fn update_record(&mut self, id: &str, new_record: Record) {
````
as mentioned before, this might be a good method to have, but we may also want to have `update_sequence()`? So that no new allocation of records is necessary when updating just the sequence itself, nothing else about it

MattesMrzik replied:
I think records are not modifiable: https://docs.rs/bio/latest/bio/io/fasta/struct.Record.html

junniest commented on phylo/src/alignment/tests.rs:
```diff
@@ -474,3 +474,79 @@ fn sequence_iterator_access() {
+
+#[test]
+#[should_panic]
+fn update_ancestral_maps_fails_on_leaf() {
```
I think that here it would be helpful to add a comment as to why exactly it panics? I.e. what condition does this new map not satisfy.

junniest replied:
Ah, you're not supposed to be able to set a map on the leaf, only internal nodes? I didn't process the name properly. Then it is probably fine the way it is.

junniest commented on phylo/src/alignment/tests.rs:
```diff
@@ -474,3 +474,79 @@ fn sequence_iterator_access() {
+    let new_map = align!(b"A-AA-AAA-A");
+    let node_id = "A1";
+    let node_idx = &tree.idx(node_id);
+    msa.update_ancestral_map(node_idx, new_map.clone());
```
nitpick: unnecessary `clone()`

junniest commented on phylo/src/alignment/tests.rs:
```diff
@@ -474,3 +474,79 @@ fn sequence_iterator_access() {
+    let new_map = align!(b"A-AAAA-A");
+    let node_id = "I3";
+    let node_idx = &tree.idx(node_id);
+    msa.update_ancestral_map(node_idx, new_map.clone());
```
nitpick: unnecessary `clone()`

junniest commented on phylo/src/optimisers/nni_optimiser.rs:
```diff
@@ -50,6 +50,7 @@ impl MoveOptimiser for NniOptimiser {
         let mut max_cost = f64::MIN;
         for child_idx in &cost.tree().node(node_idx).children {
             // TODO: parallelization?
+            // TODO: can we avoid cloning the whole cost.model_info?
```
probably not if any parallelisation is (potentially) involved, since these are the temp values that will change if anything in the model/tree changes...

junniest replied:
Unless there are some intermediate values that will never be written, only read, those could be a shallow copy (not clone), but it might be tricky to separate them out.

MattesMrzik replied:
I see, here its necessary, but perhaps we could return the cost `C` and not just the cost `f64`.
See #141

junniest commented on phylo/src/optimisers/topo_optimiser.rs:
```diff
@@ -209,6 +209,10 @@ where
 
                 if best_cost > base_cost {
+                    // TODO: can we not re-use the cost struct that was used to compute the
+                    // best_move_at_location, instead of calling update_tree() again?
```
That may be a good idea. Instead of trying to update the cost entity here, replace it with the final one from the move.
Would that mean that the signature of 
` fn single_optimisation_iteration(&mut self) -> Result<f64>`
 will change to something like 
`fn single_optimisation_iteration(&mut self) -> Result<C>`?

junniest replied:
And probably the rest of places that return `f64` or `MoveCostInfo` should return the cost function instead. I have a feeling it might become confusing, since sometimes we shouldn't update the cost at all, if no better move is found. But that would remove the need to exactly copy the move-introduced changes into the old cost function. I'm fairly certain that the whole best move selection process should be refactored though to untangle what is updated when...

MattesMrzik replied:
>Would that mean that the signature of
 fn single_optimisation_iteration(&mut self) -> Result<f64>
will change to something like
fn single_optimisation_iteration(&mut self) -> Result<C>?

I dont think so, since it is `&mut self`. However the `best_move_at_location` in `fold_improving_moves` would benefit from reusing the cost (especially if TKF is used, since ancestral seqs dont need to be re-estimated twice).

MattesMrzik replied:
Lets continue this discussion in #141

junniest commented on phylo/src/optimisers/topo_optimiser_tests.rs:
```diff
@@ -937,3 +938,54 @@ fn example_main_from_readme() {
+    let msa = dir.join("masa.fasta");
+    let tree = dir.join("tree.newick");
+    let phylo = PIB::with_attrs(msa, tree).build_with_ancestors().unwrap();
+    phylo.check_dollos_constraint().unwrap();
```
Is this check necessary here? I think it belongs to a separate unit test, since the data you set up here wouldn't make sense if the test fails.

junniest commented on phylo/src/phylo_info/mod.rs:
```diff
@@ -147,6 +150,52 @@ impl<A: Alignment> PhyloInfo<A> {
+    /// i.e., each alignment column must be introduced exactly once
+    /// in the phylogenetic tree.
+    ///
+    /// Returns `Ok(())` if all columns satisfy this constraint, or an error
```
Hmmm. Using `Result` like that seems a bit odd to me, but I wouldn't know how to make it better. One suggestion from gemini is to use a properly defined Error, e.g. something like `DolloViolation(Vec<(usize, usize)>)` to make sure there's a way to read the indices without parsing strings. However, the PR with crate-specific errors is still in progress and I haven't managed to merge/finalise it, so I think it's ok for now.

junniest replied:
The reasoning for the change is that non-specific error handling such as what anyhow provides is only meant for binaries (something I didn't know when setting up the lib). The libraries should define their own specific error types that can then be mixed all together with anyhow in binaries.

MattesMrzik replied:
See #129

junniest commented on phylo/src/tkf_model/tkf91.rs:
```diff
@@ -29,6 +29,15 @@ pub struct TKF91IndelModel {
 
+impl TKF91IndelModel {
+    #[cfg(test)]
+    pub(super) fn default() -> Self {
```
I think that this `default()` which is now a method should become an impl of the Default trait. It can still be useful for other devs to get a starting model with valid parameters -- especially when initialising model search. I generally feel that we should not use trait method names such as iter and default without implementing the traits, but also one of the Rust api recommendation is to implement these basic traits anyway wherever meaningful.

junniest commented on phylo/src/tkf_model/tkf92.rs:
```diff
@@ -80,6 +92,8 @@ impl TKFModel for TKF92IndelModel {
     }
 
+    // this is not a prob but a factor since it can be > 1
+    // TODO i really should rename also the node_event_prob to node_event_factor or something
```
lol, please don't forget to address the todo's, or transform them into issues so that we don't lose track! There's quite a few todos strewn around.

MattesMrzik replied:
before merge i will go through and collect them

junniest commented on phylo/src/tkf_model/tkf92.rs:
```diff
@@ -236,7 +228,34 @@ impl<Q: QMatrix, AA: AncestralAlignment> TKF92CostBuilder<Q, AA> {
+/// Determines the block borders from the alignment. A block border is defined as a
+/// position where any sequence changes from gap to non-gap or vice versa. Returns a sorted
+/// vector of the right exclusive block borders.
+pub(crate) fn blocks_of_alignment<AA: AncestralAlignment>(msa: &AA) -> Vec<usize> {
```
Would it maybe be helpful to return a struct that contains both the block starting positions and lengths? I know that you need to get the lengths somewhere down the line too, but I cannot find the place now. This would mean that you don't have to go through the blocks twice, since you should already know the lengths here.

MattesMrzik replied:
In `blocks_of_alignment` at the end i am sorting the list of block borders. This needs to be done. Only after that i can determine the block lengths. It might be a bit nicer to have a struct of blocks, but then iterating all blocks becomes more complicated since we would have to borrow the blocks from the model_info. But in the loop we again need a mutable access to model_info. So before every loop we would have to clone these instead of just doing `x in range 0..number_of_blocks`

junniest replied:
I see. Fair enough!

junniest commented on phylo/src/tkf_model/tkf92_additional_blocks.rs:
```diff
@@ -0,0 +1,256 @@
+/// which means that the provided blocks will be used in addition to the blocks determined from
+/// the alignment, see [`super::TKFModel::get_blocks`].
+#[derive(Clone, Debug, PartialEq)]
+pub struct TKF92IndelModelAddBlocks {
```
Damn there's a lot going on here. Why is this struct necessary? I don't think I can figure this out just by reading the code. It also looks like quite a bit of code duplication in the different TKF variants...

MattesMrzik replied:
This is indeed much duplication.
I currently use this struct only for testing. But i thought if I implemented it why not make it available to others. But there might be no reason why anybody would use it.

MattesMrzik replied:
Will mark them as `#[cfg(test)]`

junniest replied:
Fair, yes, I did not realise it's something for testing only. Could it also be moved into a test module?

MattesMrzik replied:
I would like to have it in its separate file, but giving the file name a suffix like `_test`? Or place it in a dir `test_costs`'?

junniest replied:
I'm still not sure how and when this is used. Does it need any internals from TKF models or could this be moved to a separate test dir (on the same level as src) to become part of integration testing?

MattesMrzik replied:
It is used to check if the likelihood as it is returned by the DP reestimation is the same as if the logl is calculated from a fresh cost struct. Since the blocking as inferred from the alignment may change during reestimation we need to create the new cost with modified (ie additinal) blocks. Therefore, you can see the two use cases in the tests: `tkf92_reestimate_large_tree_for_file_iterative` and `tkf92_update_tree`.
Perhaps in the future when calling reestimate updates the blocking, this file is no longer needed. 

It could be moved to an integration test folder.

junniest replied:
So, while it is still used, it can get moved to a different folder, correct? Then for now you can move it into a `tests` folder on the same level as src. Rust will pick it up and run it as a regular test (but it won't mess with the runtimes of unit tests) (https://doc.rust-lang.org/rust-by-example/testing/integration_testing.html)

MattesMrzik replied:
mmh, i am using this `TKF92IndelModelAddBlocks` in reestimation and in topology optimization. So i would keep this file (for now) in the crate, see #154. The tests however I will move to the integration `tests` dir. 
Then however, `TKF92IndelModelAddBlocks` can not be conditionally compiled only when #[cfg(test)].

junniest commented on phylo/src/tkf_model/tkf_indel.rs:
```diff
@@ -400,54 +420,97 @@ impl<T: TKFModel, AA: AncestralAlignment> ModelSearchCost for TKFIndelCost<T, AA
+        self.phylo.tree = tree;
+        if update_due_to_nni {
+            let v2 = self.tree().nodes[dirty_nodes[0]].idx;
+            let rng = &mut DefaultGenerator::default();
```
We should be passing the rng on initialisation, not instantiating it on the go, since this way there is absolutely no guarantee of reproducibility. The idea is that if the seed is set to a specific value, you should get the exact same result, which is not possible when the generator is created in the middle of a method.

MattesMrzik replied:
Then we have at least 3 options: 
- pass `rng` to `update_tree` in the `TreeSearchCost` (which would feel a bit weird, since "why does a simple update_... method need rng?")
- have a new trait building upon or taking inspiration from TreeSearchCost that has a method update_tree_and_ancestral_alignment which would take the rng as an argument. However this would also come with lots of refactoring, since the spr, nni, and branch lengths optis only work with treesearchcost right now. 
- have the `rng` as a field of the `cost.` Then i think this cost would have to `own` it. Or we change how the rng is stored in the topology_optimiser (ie having `shared mutable access`)

junniest replied:
Hm. We can think about it. At the very least I think this should be marked as an issue we need to address, since it feels like too much to fix now!

MattesMrzik replied:
Agreed, see: #142

junniest reviewed (commented):
This is a long one and I will definitely need to look again! It's quite difficult to track what's going on unfortunately.

MattesMrzik committed:
changes N to X in masa

MattesMrzik committed:
renamed fixed_fragments masa file

MattesMrzik committed:
rename recompute-brute-force-ancestors feature

MattesMrzik committed:
adjusted readme feature hint

MattesMrzik committed:
added comments to update_anc_maps

MattesMrzik committed:
added comment to tests

MattesMrzik committed:
removed unnecessary clone

MattesMrzik committed:
removed dollo check on initial masa in top opti test

MattesMrzik committed:
fixed and add blocks are now test only

MattesMrzik committed:
corrected file path in re-estimate test

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -0,0 +1,966 @@
+        }
+        let argmax = if argmaxes.is_empty() {
+            debug_assert!(max == f64::NEG_INFINITY);
+            42 // dummy value, will get ignored anyway since max is -inf
```
While I understand your comment, I think this should return Ok((f64, usize)). Then no checks for specific values will be necessary since this branch will return None.

MattesMrzik replied:
Thanks! Implemented your suggestion

junniest reviewed (commented):
I have a feeling I somehow missed the `reestimate.rs` file entirely last time. I will go through it more thoroughly tomorrow.

MattesMrzik committed:
renamed features to be more clear

MattesMrzik committed:
memory allocation in update ancestral map unchecked

MattesMrzik committed:
max_over_previous returns Option instead of dummy

MattesMrzik committed:
added mentions to github issues

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -0,0 +1,966 @@
+const DP_ASSIGNMENT_AND_EVENTS_SIZE: usize = 128;
+const BACKTRACKING_INVALID: usize = DP_ASSIGNMENT_AND_EVENTS_SIZE + 1;
+/// #{v1, v2, t2, t3, t4}, see [`QuartetEdges`].
+const N_EDGES_IN_QUARTET: usize = 5;
```
nitpick: This block should probably come before the type definitions so that you use the consts you already defined.

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -0,0 +1,966 @@
+    fn t1_has_char(&self, site: usize) -> bool {
+        // in case the t1_mapping is None, v1 is the root and t1 does not exist
+        // therefore it cannot have a character
+        match &self.t1_mapping {
```
nitpick: a more idiomatic version of this (apparently) would be to use `.is_some_and()`:
```
        self.t1_mapping
            .as_ref()
            .is_some_and(|mapping| mapping[site].is_some())
```
Saves a single line of code, but looks very rusty :D

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -0,0 +1,966 @@
+/// // The alignment below includes ancestral sequences for which the indel points
+/// // will be refined.
+/// let msa = "data/tkf/reestimate/masa.fasta";
+/// let phylo = PhyloInfoBuilder::new(msa).build_with_ancestors()?;
```
Hmmm. I am a bit confused about how this works. Does this inherently rely on the fact that the nj tree will match the MASA structure/nodes?.. Looking at the `build_with_ancestors()` method, it doesn't seem like it uses the internal sequences at all, it would just build an nj tree from scratch. How do the ancestors from the MASA come into play?

I'm now thinking that providing a MASA without a tree structure should not be allowed in the first place, since there isn't really a way to reconstruct the tree with matching internal node names that would make sense with the MASA...

MattesMrzik replied:
if a tree file is provided then we dont use the nj tree. In the `build_with_ancestors` we have the case`if sequences.len() == tree.len()`, so we use the tree from the file and associate the seqs from the MASA with every node (leaves and internals). 

Regarding your second question: please see the issue I just created: #144

junniest replied:
Fair! And I guess you are checking that the node IDs match in `from_aligned_with_ancestral`.

Thank you for creating an issue, it should be addressed. The IO is getting more complicated by the moment...

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -0,0 +1,966 @@
+                self.cost.logl();
+                break;
+            }
+        }
```
I am always weary of iterating through nodes, since that slows down access. I think this could be replaced by:
```
        if !self
            .cost
            .model_info
            .borrow()
            .valid_for_reestimation
            .is_full()
        {
            warn!("Reestimation can only be performed on a cost with valid_for_reestimation internal nodes tmp values. Making them valid now.");
            self.cost.logl();
        }
```
Since all of the values have to be set to true for reestimation to be valid?

junniest replied:
Ah, I guess it's just the internal node values that have to be true... But you check all nodes anyway in the preorder and you set the value to true for the leaves, so this code should be equivalent (and potentially faster).

At least if I run the reestimation tests with this version of the code they all pass (:

MattesMrzik replied:
Thanks for the hint:)

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -0,0 +1,966 @@
+    /// Resets the DP and backtracking tables. Initialises the [`QuartetEdges`]. Removes the old
+    /// quartet contributions from the root aggregated values.
+    ///
+    /// Returns an error if something goes wrong during quartet edge initialisation.
```
Hm. What are the potential errors here? This comment is good, but it doesn't say what to fix and what can be done about it. It seems like most of the operations you perform cannot fail.

As far as I can see it's only `self.quartet_edges = QuartetEdges::new(v2_idx, self.cost)?;` that can fail, and that only fails when the node is the root or doesn't have a sibling?

Since `reestimate()` already handles some error cases, I would suggest that it should handle these cases too. Actually, it does check that the node is not the root already, so that check in `QuartetEdges::new()` is redundant.

I would suggest doing all the error handling in the `reestimate()` method and then call `reestimate_unchecked()` as we do in other scenarios.

MattesMrzik replied:
I split it into `reestimate() -> Result<f64>` and `reestimate_unchecked() -> f64`. Please see the updated code:)

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -0,0 +1,966 @@
+        &self.cost.phylo
+    }
+
+    pub fn reestimate(&mut self, v2_idx: &NodeIdx) -> f64 {
```
Is there a reason why this prints a warning but still returns the likelihood? It feels counterintuitive to me, especially since the `self.logl()` should already have the same value. I think it's also confusing the logic because some of the internal calls return `Result`s, but this one handles that error internally and prints warnings instead.

Moreover, if the reestimation failed, should we proceed at all with the process? It seems to me that in this case the "failure" only happens if node v2 is unsuitable for reestimation, which just means that we shouldn't do anything, not that the reestimation failed.

I suggest refactoring this to put all error handling together and rethinking whether Result, Ok or just f64 as the return value is the best call.

MattesMrzik replied:
I split it into `reestimate() -> Result<f64>` and `reestimate_unchecked() -> f64`. Please see the updated code:)

junniest replied:
LGTM!
nitpick: the error messages should start with a lowercase letter and should not have contractions, i.e.:
```
"reestimation cannot be performed for the root '{v2_id}'."
```
and:
```
"reestimation cannot be performed for leaf node '{v2_id}'."
```

MattesMrzik replied:
And I guess also without the `.` at the end.

junniest commented on phylo/src/tkf_model/reestimate/mod.rs:
```diff
@@ -0,0 +1,966 @@
+            self.cost
+                .model_info
+                .borrow_mut()
+                .valid
+                .set(usize::from(*edge), false);
+        }
```
This unnecessarily does `borrow_mut()` 5 times which includes runtime checks to ensure safe borrowing rules. I don't think there can be any borrow conflicts here (or if there will be it's a bug), so it's much better to borrow once:
```
        let mut model_info = self.cost.model_info.borrow_mut();
        for edge in self.quartet_edges.edges() {
            model_info.valid.set(usize::from(*edge), false);
        }
```

junniest replied:
Potentially if you do `borrow()` or `borrow_mut()` multiple times in other places in a cycle for example, you should apply similar refactoring.

MattesMrzik replied:
I think some of it cant be optimized right now, since the rng is as a `    rng: &'a mut RandomGenerator<R>,
`. Therefore methods that access the `rng` have to be `&mut self` and this clashed with the refcell of the model_info. However I will take a closer look at this in issue #151.

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -0,0 +1,966 @@
+                let event = self.cost.determine_event(edge, block_id);
+                let node_event_prob = self.cost.event_prob(edge, event);
+                let node_eta = self.cost.eta_for_non_root(edge, event);
+                self.cost.update_previous_event(edge, event);
```
Hmmm. On the one hand, I really like the individual smaller functions that do their own thing. On the other, `update_previous_event` also does a mutable borrow of the same instance of `model_info`, meaning that we not only borrow within a cycle, we also borrow mutably at least twice.

Could `update_previous_event` be refactored to return the new event state instead of setting it? Although it would then need to return the previous event (i.e. read `model_info`) and we might end up setting the value to the same one as it was before. Although `update_previous_event()` could return `Ok(bool)`, which wouldn't require reading the existing value and we can use `is_none_or(|val| model_info.previous_event_deletion.set(usize::from(edge), val)` to avoid unnecessary setting.

MattesMrzik replied:
Like so?
getting the new val
```rust
    pub(super) fn updated_previous_is_deletion(&self, event: Event) -> Option<bool> {
        match event {
            Event::Deletion => Some(true),
            Event::Insertion | Event::Homolog => Some(false),
            Event::Nothing => None,
        }
    }
```
and setting the val where previously `self.cost.update_previous_event(edge, event);` was called:

```rust
if let Some(val) = self.cost.updated_previous_is_deletion(event) {
    model_info
        .previous_event_deletion
        .set(usize::from(*edge), val);
}
```

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -0,0 +1,966 @@
+        current_del_or_not: &QuartetDelOrNot,
+        current_events: &QuartetEvents,
+    ) -> QuartetDelOrNotPossibilities {
+        // let mut choices: Vec<Vec<bool>> = vec![vec![]; N_EDGES_IN_QUARTET];
```
nitpick: the commented out line should be deleted

MattesMrzik replied:
thanks for spotting it:)

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -0,0 +1,966 @@
+    /// Based on the `current_events` and `del_or_not` finds all compatible previous `del_or_not`.
+    // TODO: these are not ensured to be compatible with the previous assignments. Would it be
+    // worth to filter them here?
+    fn prev_compatible_del_or_not(
```
I am struggling with the distinction between this method and `possibilities_for_del_or_not`. What's the difference? Could they be renamed for clarity?

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -0,0 +1,966 @@
+        // go back the path
+        for block_id in (0..(n_blocks - 1)).rev() {
+            if came_from == BACKTRACKING_INVALID {
+                panic!("Backtracking table contains invalid value at block_id = {block_id}");
```
Is this reachable? Shouldn't be. I would recommend using the `unreachable!()` macro instead of `panic!` since we don't handle this case and being here would certainly be a bug.

MattesMrzik replied:
Thanks!

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -0,0 +1,966 @@
+    t2_has_char: bool,
+    t3_has_char: bool,
+    t4_has_char: bool,
+) -> Vec<(bool, bool)> {
```
The return type here should be `EdgeAssignmentPossibilities`.

junniest commented on phylo/src/tkf_model/reestimate/mod.rs:
```diff
@@ -0,0 +1,966 @@
+    }
+
+    /// Finds the maximum value in the last column of the DP table and its index.
+    /// If there are multiple maxima, one is chosen at random.
```
For now, to avoid issues with reproducibility, it might make sense to have the option to take the first max deterministically until we figure out how to pass the rng.

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -0,0 +1,966 @@
+                continue;
+            }
+            const_per_alignment +=
+                log_i1(l, self.cost.model_info.borrow_mut().beta[usize::from(node)]);
```
why is this `borrow_mut()`? I think it only needs to be `borrow()`.

MattesMrzik replied:
yep, thanks:)

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -0,0 +1,966 @@
+        let l = self.cost.model.lambda();
+        let m = self.cost.model.mu();
+        let mut const_per_alignment: f64 = (1.0 - l / m).ln();
+        for node in self.cost.phylo.tree.postorder() {
```
Why is this one `postorder()`? I'm guessing the order doesn't matter here anyway, but it is confusing why sometimes you go for preorder and sometimes for post. If you do preorder in this case you can just skip the first node and remove the check in the next line.

junniest commented on phylo/src/tkf_model/reestimate/mod.rs:
```diff
@@ -0,0 +1,966 @@
+        }
+        const_per_alignment
+    }
+}
```
With my two previous comments and to try and avoid runtime checks on the multiple borrows, I suggest the following rewrite of this:
```
    fn const_per_alignment(&self) -> f64 {
        let l = self.cost.model.lambda();
        let m = self.cost.model.mu();
        let mut const_per_alignment: f64 = (1.0 - l / m).ln();

        let nodes = self.cost.phylo.tree.preorder().iter().skip(1); // skip root node
        let info = self.cost.model_info.borrow();

        for node in nodes {
            const_per_alignment += log_i1(l, info.beta[usize::from(node)]);
        }
        const_per_alignment
    }
```

MattesMrzik replied:
Thanks!

junniest commented on phylo/src/tkf_model/reestimate.rs:
````diff
@@ -0,0 +1,966 @@
+/// # Example
+/// ```rust
+/// let no_choices = [true, false, false, true, false];
+/// let choices = vec![1, 3];
````
If I get this correctly, this is something that will never change, right? For a specific array of `no_choices` and a vec of `choices` the outcome will always be the exact same? If that is the case, I suggest pre-computing this as `lazy_static` and accessing the precomputed values instead of computing them on the fly every time. 

If this computation happens repeatedly (and I assume that it does), it will save a lot of time at the expense of some memory. It's probably possible to save memory on this as well by using a fixed size bitset for the `no_choices` var, since it's only ever going to be of length 5. 

It might be helpful to use bitsets of size 8 anyway because that can save time on access, but I don't know if that's something that rust can optimise automatically. Asking copilot for a benchmark could help here.

One more thing is that I am still struggling to understand what each of the inputs mean from the names. The `no_choices` seems like something that shouldn't change, but the positions in `choices` indicate the ones that vary. We can discuss this, but I think there's a refactor there that would help clarify what's going on.

MattesMrzik replied:
> ... this is something that will never change ...

Yes, given the input (i.e., some 5 long boolean vec and a sorted set `s \subset {0,1,2,3,4}.`, the output is always the same.

> I suggest pre-computing this as lazy_static

Yep, I can do that. 

> ... using a fixed size bitset ...

I think arrays are allocated on the Stack, while `FixedBitSet` on the heap. So it might be faster to stick to arrays.

> ... what each of the inputs mean from the names ...

I wanted to work on the stack, hence the arrays and the logic that the `no_choices` is meant to be a template on which states are defined as fixed unless they are varied (between true and false) if their idx is in `choices`. Additionally, `choices` could be renamed to `positions_to_vary` and `no_choices` to `base`.

Alternatively, I could collect over the edges in the quartet the possibilities, i.e., either [true] (for last event is a deletion), [false] (last event isnt a deletion) or [false, true] (if we dont yet know if the last event was a deletion or not). Then we could also use itertools to get all combinations and cache the results in a lazy_static, a table of of bools of approx size 64 x 5.

junniest replied:
Fair point for the arrays. 

The renaming seems meaningful to me! I like it.

And yes, I think that collecting and caching the combinations would be the best way to do this, it's a good idea to cache anything that you will need repeatedly but that will never change.

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -0,0 +1,966 @@
+
+/// Converts the provided index as used by the DP table into the corresponding `assignment` and
+/// quartet `del_or_not` combination.
+fn index_to_bools(index: usize) -> ([bool; 2], Vec<bool>) {
```
Ok, this one is very confusing. Why do you need to do bit math? And why does it return a tuple from an array of two (which should probably be a tuple) and a vec? This needs more clarification and type definitions.

MattesMrzik replied:
I now used the correct types:
```rust
/// Converts the provided index as used by the DP table into the corresponding `assignment` and
/// quartet `del_or_not` combination.
/// Is the inverse of [`bools_to_index`].
/// During backtracking we only need to know the assignment.
fn index_to_bools(index: usize) -> (EdgeAssignment, QuartetDelOrNot) {
    let mut bits = index;
    let mut q_del_or_not = [false; N_EDGES_IN_QUARTET];

    // extract del_or_not booleans first
    for i in (0..N_EDGES_IN_QUARTET).rev() {
        q_del_or_not[i] = (bits & 1) != 0;
        bits >>= 1;
    }
    // then extract assignment booleans
    let assignment = ((bits & 2) != 0, (bits & 1) != 0);

    (assignment, q_del_or_not)
}
```

junniest commented on phylo/src/tkf_model/reestimate/mod.rs:
```diff
@@ -0,0 +1,966 @@
+
+/// Converts the provided assignment and quartet `del_or_not` combination
+/// into a unique index for the DP table.
+fn bools_to_index(assignment: &EdgeAssignment, q_del_or_not: &QuartetDelOrNot) -> usize {
```
I think you need to explain the logic here. The comment says what it is supposed to do but not why or how.

MattesMrzik replied:
How do you like: 
```rust
/// Converts the provided assignment and quartet del_or_not combination
/// into a unique index for the DP table. The DP algorithm calculates probabilities
/// for such an assignment and quartet del_or_not combination. To store these     
/// results in a flat array, we need to convert the combination of booleans
/// into a unique index.
fn bools_to_index(assignment: &EdgeAssignment, q_del_or_not: &QuartetDelOrNot) -> usize {
    // Iterate over all booleans (i.e., assignment and del_or_not concatenated):
    // first shift the index to the left by 1 (multiply by 2)
    // then add 1 if the boolean is true, or 0 if it is false
    [assignment.0, assignment.1]
        .iter()
        .chain(q_del_or_not.iter())
        .fold(0, |index, &b| (index << 1) | (b as usize))
}
```

junniest commented on phylo/src/tkf_model/reestimate/mod.rs:
```diff
@@ -0,0 +1,966 @@
+    (assignment, event)
+}
+
+fn event_for_edge(node_has_char: bool, parent_has_char: bool) -> Event {
```
This should probably be inline? And/or static?

MattesMrzik replied:
Added the `#[inline]`.  By static you mean a static lookup table? I think this will not make a noticeable difference in runtime. But if you disagree, I can apply the same procedure as for `possible_assignments_of_edge`:)

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -0,0 +1,966 @@
+    }
+}
+
+pub(super) fn possible_assignments_of_edge(
```
I'm guessing that this function could be a lookup table instead, since this isn't something that will change through time. E.g. the combination of `false, false, false, false` (can be represented in a bitset of `0000`) will always result in `vec![(false, false)]`. You could use lazy_static to precompute this at compile time and use for lookup.

I understand that this is a relatively small amount of computation, but if it is called repeatedly, it will add up.

MattesMrzik replied:
I think lazy static is in runtime. I have modified to a static lookup table: 
```rust 
const DOLLO_ASSIGNMENTS: [&[(bool, bool)]; 16] = [
    /* 0000 */ &[(false, false)],
    /* 0001 */ &[(true, true), (false, true), (false, false)],
    /* 0010 */ &[(true, true), (false, true), (false, false)],
    /* 0011 */ &[(true, true), (false, true)],
    /* 0100 */ &[(true, true), (true, false), (false, false)],
    /* 0101 */ &[(true, true)],
    /* 0110 */ &[(true, true)],
    /* 0111 */ &[(true, true)],
    /* 1000 */ &[(true, true), (true, false), (false, false)],
    /* 1001 */ &[(true, true)],
    /* 1010 */ &[(true, true)],
    /* 1011 */ &[(true, true)],
    /* 1100 */ &[(true, true), (true, false)],
    /* 1101 */ &[(true, true)],
    /* 1110 */ &[(true, true)],
    /* 1111 */ &[(true, true)],
];

pub(super) fn possible_assignments_of_edge(
    t1_has_char: bool,
    t2_has_char: bool,
    t3_has_char: bool,
    t4_has_char: bool,
) -> Vec<(bool, bool)> {
    let idx = (t1_has_char as usize) << 3
        | (t2_has_char as usize) << 2
        | (t3_has_char as usize) << 1
        | (t4_has_char as usize);

    DOLLO_ASSIGNMENTS[idx].to_vec()
}

```

junniest replied:
Ah, you're correct, of course `lazy_static` it's runtime. At least it does execute only once. There's now an standard library [`std::sync::OnceLock`](https://doc.rust-lang.org/std/sync/struct.OnceLock.html) that can be used to the same effect, but will need more refactoring and I think we should stick with `lazy_static` for now unless it becomes critical.

I like this version a lot better. It does involve quite some type conversions, but I think it will be faster than the original implementation. (Here I went into a rabbit hole of reading about type conversions in Rust, which just told me that converting bool to u8 and bool to usize will have the same performance and I should stop overthinking.)

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -0,0 +1,966 @@
+
+    /// Returns all possible combinations of `deletion or not` for each edge in the quartet
+    /// given the (current) events taken on those edges.
+    fn possible_del_or_not(
```
I am struggling with the distinction between this method and `possibilities_for_del_or_not`. What's the difference? Could they be renamed for clarity?

MattesMrzik replied:
I renamed `possible_del_or_not` to `possible_del_or_not_for_event` and
`possibilities_for_del_or_not` to `del_or_not_combinations`
I hope this and the doc strings make this difference clearer.

junniest commented on phylo/src/tkf_model/tkf92_fixed_fragmentation.rs:
```diff
@@ -0,0 +1,448 @@
+        let merged = merge_fragmentation_with_blocks(&fragmentation, &blocks);
+        assert_eq!(merged, blocks.clone());
+        assert_eq!(merged, naive_merge(&fragmentation, &blocks));
+    }
```
These tests can be merged into 1 using `#[rstest]`, see e.g. `phylo/src/parsimony/scoring/tests.rs`

MattesMrzik replied:
thats true, thanks:)

junniest commented on phylo/src/tkf_model/tkf92_fixed_fragmentation.rs:
```diff
@@ -0,0 +1,448 @@
+    let mut fragmentation = fragmentation.to_vec();
+    let original_len = fragmentation.len();
+    if original_len == 0 {
+        return fragmentation.to_vec();
```
nitpick: you could probably return `vec![]` here to avoid transformation and make it clearer that the return vector is empty.

junniest reviewed (commented):
Got quite some comments on the `reestimate.rs`. I understand that it contains complex logic, but I think it's worth trying to refactor a bit and precompute a bunch of values to make it more obvious what's going on.

MattesMrzik committed:
implemented simple requested changes

MattesMrzik committed:
implemented more of the suggestions

MattesMrzik committed:
updated rust->toml to 1.88. fixed clippy warn

MattesMrzik committed:
workflow files use rust 1.88

MattesMrzik committed:
fixed clippy

MattesMrzik committed:
max_of_last_col fixed, empty alignment blocks fix
- more tests

github-actions[bot] committed:
Format Rust code using rustfmt

MattesMrzik committed:
fmt

MattesMrzik committed:
Merge remote-tracking branch 'upstream/develop' into 115-TKF-tree-seach

MattesMrzik committed:
Merge branch '115-TKF-tree-seach' of github.com:MattesMrzik/rust-phylo into 115-TKF-tree-seach

MattesMrzik committed:
fixed doc test. cited github issues at todos, renamed prob to factor

MattesMrzik committed:
optimized: less refcell.borrow

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -210,46 +222,73 @@ where
+    ///
+    /// # Errors
+    /// Reestimation is only defined for non-root internal nodes that have
+    /// a sibling. Accordingly, this method will return an error if this is not the case.
```
Hm. I think this phrasing is confusing. Since our trees are binary, all non-root nodes always have siblings. Plus in the end the only thing that matters is that the node has a parent and isn't a leaf, so the sibling condition is never checked. I suggest removing the `that have a sibling` part from the comment or alternatively saying that we need the tree to always be binary. Although it is binary as per implementation, which this is tied to anyway.

MattesMrzik replied:
Yep, at the time of writhing that doc i wasnt sure if our trees are always binary. I will adjust this commet.

 I added a comment to the `from_newick` function saying that only binary trees are supported. Perhaps a comment to the Tree struct itself would also be appropriate?

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -275,52 +311,51 @@ where
-    /// Returns an error if something goes wrong during quartet edge initialisation.
-    fn prepare_for_dp(&mut self, v2_idx: &NodeIdx) -> Result<()> {
+    /// # Panics
+    /// Panics if `v2_idx` is the root or has no sibling.
```
Again, this is impossible given the structure, and the main requirement is that the node should have a parent.

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -351,13 +386,18 @@ where
+                let node_event_factor = self.cost.event_factor(edge, event);
                 let node_eta = self.cost.eta_for_non_root(edge, event);
-                self.cost.update_previous_event(edge, event);
+                // self.cost.update_previous_event(edge, event);
```
nitpick: haha, another commented line that should go

junniest reviewed (approved):
LGTM!

MattesMrzik committed:
cleanup

MattesMrzik committed:
more caching in reestimator

MattesMrzik committed:
cleanup for del or not combi table

MattesMrzik committed:
Merge remote-tracking branch 'upstream/develop' into 115-TKF-tree-seach

MattesMrzik committed:
impled crate::Error for reestimator, fixed bug
- the bug from the merge was addition of check if node is root in determine_event and
  then failing, even though it shouldnt

junniest commented on phylo/src/phylo_info/mod.rs:
Outdated comment, since we now use our crate's `bail!` macro

MattesMrzik replied:
true, thanks:)

junniest commented on phylo/src/tkf_model/brute_force_ancestors_tests.rs:
```diff
@@ -151,8 +151,15 @@ fn cost_for_edge_seqs<T: TKFModel>(
     let (new_v1_mapping, new_v2_mapping) = edge_seqs_to_mappings(edge_seqs, &block_lens, seq_len);
-    cost.phylo.msa.update_ancestral_map(v1_idx, new_v1_mapping);
-    cost.phylo.msa.update_ancestral_map(v2_idx, new_v2_mapping);
+    // The expect() are never get triggered, unless something is seriously wrong with the algo.
```
nitpick: grammar, either "are" or "get" should go

junniest commented on phylo/src/tkf_model/mod.rs:
```diff
@@ -101,9 +101,15 @@ impl<Q: QMatrix, T: TKFModel, AA: AncestralAlignment> TreeSearchCost for TKFCost
 }
 
 impl<Q: QMatrix, T: TKFModel, AA: AncestralAlignment> TKFCost<Q, T, AA> {
+    /// Returns a reference to the multiple ancestral sequence alignment.
```
question: I wonder if these two pub methods need doctest examples?..

MattesMrzik replied:
I added some like so: 
```rust
    /// # Example
    /// ```rust
    /// use phylo::phylo_info::PhyloInfo;
    /// use phylo::substitution_models::{SubstModel, dna_models::GTR};
    /// use phylo::alignment::{Alignment, AncestralAlignment, Sequences, MASA};
    /// use phylo::tkf_model::TKF92CostBuilder;
    /// use phylo::{tree, record_wo_desc as record};
    /// # use phylo::Result;
    /// # fn main() -> Result<()> {
    /// let tree = tree!("(((A1:2.0,B2:2.0)I3:0.3,C4:2.0)R5:1.0);");
    /// let msa = MASA::from_aligned_with_ancestral(
    ///     Sequences::new(
    ///         vec![
    ///             record!("A1", b"--GTGGA---"),
    ///             record!("B2", b"-------NNA"),
    ///             record!("I3", b"--T-------"),
    ///             record!("C4", b"AGG-------"),
    ///             record!("R5", b"--A-------"),
    ///            ],
    ///     ),
    ///   &tree,
    /// )?;
    /// let phylo = PhyloInfo { msa, tree };
    /// let subst_model = SubstModel::<GTR>::new(&[], &[]);
    /// let cost = TKF92CostBuilder::new(0.4, 0.5, 0.8, subst_model, phylo).build()?;
    /// assert_eq!(cost.masa().seqs().len(), 3);
    /// assert_eq!(cost.masa().ancestral_seqs().len(), 2);
    /// # Ok(()) }

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -36,6 +39,19 @@ const DOLLO_ASSIGNMENTS: [&[EdgeAssignment]; 16] = [
 ];
+/// The number of bits used to encode a single `del_or_not` possibility in the [`DEL_OR_NOT_TABLE`].
+const ENCODING_SIZE: usize = 2;
+const CAN_BE_VARIED_ENCODING: usize = 0b011;
```
nitpick: could be renamed to `VARIED_ENCODING` for conciseness? Or `VARIABLE`.

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -36,6 +39,19 @@ const DOLLO_ASSIGNMENTS: [&[EdgeAssignment]; 16] = [
+const ENCODING_SIZE: usize = 2;
+const CAN_BE_VARIED_ENCODING: usize = 0b011;
+const DELETION_ENCODING: usize = 0b1;
+const NO_DELETION_ENCODING: usize = 0b0;
```
nitpick: could these just be called `VARIED`/`VARIABLE`, `DELETION` and `NO_DELETION`? Or something like `DELETION_CODE`, just to make it a bit shorter.

MattesMrzik replied:
I like to have the `..._CODE`

junniest commented on phylo/src/tkf_model/mod.rs:
```diff
@@ -101,9 +101,15 @@ impl<Q: QMatrix, T: TKFModel, AA: AncestralAlignment> TreeSearchCost for TKFCost
 }
 
 impl<Q: QMatrix, T: TKFModel, AA: AncestralAlignment> TKFCost<Q, T, AA> {
+    /// Returns a reference to the multiple ancestral sequence alignment.
```
question: I wonder if this and the following methods need examples?

MattesMrzik replied:
see above

junniest commented on phylo/src/tree/mod.rs:
```diff
@@ -59,6 +58,7 @@ impl From<NodeIdx> for usize {
     }
 }
 
+/// A binary phylogenetic tree, rooted or unrooted.
```
Actually, we always end up with a rooted tree, even when we parse an unrooted newick string (see `tree/tree_parser.rs:135`). It's for algorithmic convenience (and we could output unrooted trees if necessary, though I'd have to add that funcitonality), but at the moment we cannot internally store unrooted trees.

MattesMrzik replied:
Thanks for clarification, will adjust this comment

junniest commented on phylo/src/tkf_model/tkf_indel.rs:
```diff
@@ -345,8 +336,9 @@ impl<T: TKFModel, AA: AncestralAlignment> TKFIndelCost<T, AA> {
                 _ => unreachable!("The parent of a node cannot be a leaf."),
             },
-            None => true, // root has no parent
+            None => true, // root has no parent, so we can consider it as a gap
```
nitpick: we can consider it, or we just do? I suggest a minor rephrasing to `root has no parent, so we treat the position as a gap`

A more philosophical question here is whether it's a gap or a deletion event, since you are updating events, not characters here. Another philosophical question is whether it should be considered a deletion at the root (which doesn't make much sense since it should have never existed in the first place) or an insertion later in the tree. But this is just me thinking out loud, no changes required (; Terminology here is frustratingly difficult to keep consistent.

MattesMrzik replied:
thanks. i clarified the comment a bit.

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -701,49 +726,78 @@ where
+fn events_to_idx(events: &QuartetEvents) -> usize {
+    let mut idx = 0;
+    for event in events {
+        let can_be_varied = matches!(event, Event::Nothing); // we can't vary if there is an event
```
nitpick: Would this be better described as a `match` clause instead of several `matches!` macros and if clauses?

MattesMrzik replied:
Yes, indeed. Adjusted it.

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -1237,4 +1333,176 @@ mod private_tests {
+    }
+
+    #[test]
+    fn tkf_possible_del_or_not_table_idx() {
```
I struggle to process this test. It seems like it's doing too many things at once. Would it be possible to split this into 2 or more smaller test scenarios?

MattesMrzik replied:
I did some cleanup which reduced size and improved readability.

junniest commented on phylo/src/tkf_model/reestimate.rs:
```diff
@@ -36,6 +39,19 @@ const DOLLO_ASSIGNMENTS: [&[EdgeAssignment]; 16] = [
+/// A constant table that contains precomputed values for 'del_or_not' combinations that can be
+/// queried with [`EdgeSeqsReestimator::prev_compatible_del_or_not_table_idx`] and
+/// [`EdgeSeqsReestimator::possible_del_or_not_table_idx`].
+    pub static ref DEL_OR_NOT_TABLE: [QuartetDelOrNotPossibilities; DEL_OR_NOT_TABLE_SIZE] =
```
I think that having this as a "compute once" table is great. My main issue with this file now is that it's very long (:D) and the methods are a bit mixed together, which makes it difficult to track logic.

I would suggest moving the `lazy_static` definitions and their associated methods into a new file (something similar to `protein_generics.rs` in `substitution_models/protein_models`). Then you also won't need comments like "Only called by `possible_del_or_not_table` during generation of [`DEL_OR_NOT_TABLE`].", since these functions will be cleanly separated from the main code. You can then add a use statement just for `DEL_OR_NOT_TABLE` and the indexing functions, which would also mean that if you use the other functions by accident it will become clear.

It might also be that `DEL_OR_NOT_TABLE` should be `pub(crate)` or `pub(super)` rather than `pub`, since I am not sure how external piece of code would be able to handle that structure (and whether they might at all need to).

MattesMrzik replied:
Yes: `pub(super)`, thanks for spotting it

junniest reviewed (commented):
LGTM, but I think that `reestimate.rs` should be split up into multiple files to separate the precomputed internals from the main functionality.

junniest commented on phylo/src/tkf_model/reestimate_tests.rs:
```diff
@@ -0,0 +1,75 @@
+    let mut reestimator = EdgeSeqsReestimator::new(&mut cost, rng);
+    let root_idx = reestimator.cost.phylo.tree.root;
+
+    let err_msg = reestimator.reestimate(&root_idx).unwrap_err().to_string();
```
Ah, I missed this, but now that we have proper Errors, this should use `assert_matches!`

junniest commented on phylo/src/tkf_model/reestimate_tests.rs:
```diff
@@ -0,0 +1,75 @@
+    let mut reestimator = EdgeSeqsReestimator::new(&mut cost, rng);
+    let leaf_idx = reestimator.cost.phylo.tree.by_id("A1").idx;
+
+    let err_msg = reestimator.reestimate(&leaf_idx).unwrap_err().to_string();
```
Same here, `assert_matches!` instead of `unwrap_err`

MattesMrzik committed:
implemented most suggestions

MattesMrzik committed:
moved reestimate to dir

MattesMrzik committed:
new reestimate mod with several files
- mod.rs, tests.rs and cache.rs

MattesMrzik committed:
Merge remote-tracking branch 'upstream/develop' into 115-TKF-tree-seach

MattesMrzik committed:
fixed function call, bc fns had identical names

MattesMrzik committed:
made table pub super

MattesMrzik committed:
fixed doc string and test

junniest commented on phylo/src/tkf_model/reestimate/cache.rs:
```diff
@@ -0,0 +1,536 @@
+pub(super) const ENCODING_SIZE: usize = 2;
+pub(super) const VARIABLE_CODE: usize = 0b11;
+pub(super) const DELETION_CODE: usize = 0b01;
+pub(super) const NO_DELETION_CODE: usize = 0b00;
```
This just crossed my mind: would it maybe make more sense to have this as an Enum(usize) with a `#[repr(transparent)]` annotation?.. Just to make sure no other usize can sneak its way into the mix and ruin everything.

Another thought is this -- it now feels like you are using many different types (these usize codes, the Event Enum and booleans in other places) to describe the same thing or very similar things. Is it really necessary to have all these variants mixed together?

I don't think this has to change in this PR, but we should think about what we use to represent all the events and why.

MattesMrzik replied:
mh. the `VARIABLE_CODE` is not associated to an event. Since in `prev_compatible_del_or_not_table_idx` it's used when there is either a deletion/homolog/insertion. and in `possible_del_or_not_table_idx` it's used when we have a `Nothing` event.

junniest replied:
Hm. I think this is one of the reasons why I find it problematic -- that it's used differently in different contexts. 

Could it still be mapped to the Event enum? Event::Variable could be a new variant. Again, I am just trying to wrap my mind around all the different events/encodings you use in your code.

MattesMrzik replied:
I want the event to actually mean an event that can happen on a edge. But in the methods that get all conforming del_or_not (which is a dimension in the dp table) events are in this "weird" relationship because in the one method we look at the current del_or_not and in the other we look at previous compatible del_or_not.

junniest replied:
I see. We can always refactor later if we come up with a cleaner solution!

junniest commented on phylo/src/tkf_model/reestimate/cache.rs:
```diff
@@ -0,0 +1,536 @@
+];
+
+#[inline]
+pub(in crate::tkf_model) fn possible_assignments_of_edge(
```
I've actually never seen this annotation. In this case, what is the difference between pub(super) and this one? The variety in annotations is also bugging me a bit, I wonder if we could clean it up...

MattesMrzik replied:
I am using this in the `reestimate` module where it is also defined, but its also used in `brute_force_reestimate_tests.rs` which is outside the scope of pub(super). But we only need it in the tkf_model module and not the whole crate. therefore this visibility is used.

junniest replied:
Like I mentioned in teams, would maybe `brute_force_reestimate_tests.rs` be rather part of the `reestimate` module then?

MattesMrzik replied:
Yes, for sure!

MattesMrzik committed:
moved brute force to reestimate

MattesMrzik committed:
brute force to reestimate mod, cleanup visibility

junniest commented:
> Closes: #115, including #122. Still need to do #123

Is the description still accurate?

junniest commented on phylo/src/alignment/sequences.rs:
```diff
@@ -175,6 +175,33 @@ impl Sequences {
+            .s
+            .iter()
+            .position(|r| r.id() == id)
+            .unwrap_or_else(|| panic!("Sequence with id {id} not found"));
```
Maybe this panic! should be a bail!(Sequence, ...)?

MattesMrzik replied:
As this fn is not a change from this PR I will create a new issue for this, see  #155

junniest commented on phylo/src/tkf_model/reestimate/brute_force_ancestors_tests.rs:
```diff
@@ -0,0 +1,494 @@
+    let mut rng = FakeGenerator::default();
+    let mut reestimator = EdgeSeqsReestimator::new(&mut reestimator_cost, &mut rng);
+
+    // iterating over nodes in random order multiple times
```
Why is this necessary? Also this won't be truly random, and probably shouldn't be...

MattesMrzik replied:
Not really, i just thought that this way the test would be more thorough. removed it now

junniest commented on phylo/src/alignment/sequences.rs:
```diff
@@ -175,6 +175,33 @@ impl Sequences {
+    /// Replaces the record with the given ID with a new record.
+    ///
+    /// # Panics
+    ///
```
See below, but I think this shouldn't be a manual panic, should likely be a bail

junniest commented on phylo/src/tkf_model/reestimate/cache.rs:
```diff
@@ -0,0 +1,536 @@
+}
+
+// TODO: these are not ensured to be compatible with the previous assignments. Would it be
+// worth to filter them here?
```
It's probably worth filtering, but likely not here, but maybe at the caller?

MattesMrzik replied:
Reposted this comment to the respective issue, see #151

junniest commented on phylo/src/tkf_model/reestimate/cache.rs:
```diff
@@ -0,0 +1,536 @@
+            // since the flag 'deletion or not' does not differentiate between
+            // Insertion and Homolog events for example.
+            // In that case the mapping 'possible_del_or_not_for_event' should also be the
+            // same.
```
nitpick: Line breaks a bit weird here :D

MattesMrzik replied:
fixed it

junniest commented on phylo/src/tkf_model/reestimate/mod.rs:
```diff
@@ -0,0 +1,1014 @@
+///         let new_cost = reestimator.reestimate(&node)?;
+///         println!("Re-estimated sequences at node {node}, cost after re-estimation: {new_cost}",);
+///     }
+///     break; // stopping early for doc test, remove in real usage
```
Could you maybe hide this line from the docs using `#`? This would also mean you don't need the "remove in real usage" part of the comment because no dev will ever read the comment that far :D

MattesMrzik replied:
Did comment it. But (see comment below) if we dont have this as pub, we wouldnt need this docstring

junniest replied:
Fair, thanks!

junniest commented on phylo/src/tkf_model/reestimate/mod.rs:
```diff
@@ -0,0 +1,1014 @@
+    /// Creates a new [`EdgeSeqsReestimator`] for the provided [`TKFIndelCost`].
+    /// The reestimator can then be repeatedly used to [re-estimate](`EdgeSeqsReestimator::reestimate`)
+    /// ancestral wild card sequences for different internal node pairs.
+    pub fn new(
```
question: should this really be `pub` or should it be `pub(crate)`? Is this meaningful to call outside of the crate at all?

MattesMrzik replied:
It could be used by someone using the lib if he/she wants to reestimate the MASA. Alternatively, we could only expose this reestimation of ancestors through a routine that is similar to a optimizer, ie it iterates through nodes until no further improvement.

junniest replied:
That's a good question. I guess we can leave it as is for now and then think about including an ASR optimiser. It feels like the interface for this is a bit clunky at the moment, with the user having to iterate over internal nodes specifically.

The thing that bugs me a lot about the doctest is that the tree you use to get the list of nodes is not the "same" tree as is in the reestimator. I understand that it hasn't changed between the calls, but if someone copies the code they can add stuff and mess the whole process up.

Actually, now that I'm looking at it, this piece of code will fail when not run as a doctest (without the break). It will eventually try to reestimate at the root.

I suggest this version:
```
/// # use phylo::Result;
/// # fn main() -> Result<()> {
/// use phylo::phylo_info::PhyloInfoBuilder;
/// use phylo::random::DefaultGenerator;
/// use phylo::tkf_model::{EdgeSeqsReestimator, TKF92IndelCostBuilder};
/// use phylo::tree::NodeIdx::{Internal, Leaf};
/// // Re-estimation for ASR refinement.
/// // The alignment below includes ancestral sequences for which the indel points
/// // will be refined.
/// let tree = "data/tkf/reestimate/tree.newick";
/// let msa = "data/tkf/reestimate/masa.fasta";
/// let phylo = PhyloInfoBuilder::with_attrs(msa, tree).build_with_ancestors()?;
/// let lambda = 0.9;
/// let mu = 1.0;
/// let r = 0.5;
/// let mut tkf92_indel_cost = TKF92IndelCostBuilder::new(lambda, mu, r, phylo)
///     .build()?;
/// let mut rng = DefaultGenerator::default();
/// let mut reestimator = EdgeSeqsReestimator::new(&mut tkf92_indel_cost, &mut rng);
///
/// for node in reestimator.phylo().tree.postorder().clone() {
///     if node == reestimator.phylo().tree.root {
///         continue;
///     }
///     match node {
///         Internal(_) => {
///             let new_cost = reestimator.reestimate(&node)?;
///             println!("Re-estimated sequences at node {node}, cost after re-estimation: {new_cost}",);
///         }
///         Leaf(_) => {}
///     }
/// }
/// println!(
///     "The re-estimated ancestral MSA is {}",
///     reestimator.phylo().msa
/// );
/// println!("Re-estimation complete.");
/// # Ok(()) }
```

junniest replied:
As a way to simplify this you could add a method like `reestimate_all()` to the reestimator and it would handle walking over all the nodes internally. Then an example for `reestimate(&node)` would only need to do one internal node to illustrate reestimation.

junniest replied:
I also suggest removing the `break;` entirely. This is a pretty quick test anyway, and the break was hiding a problem.

MattesMrzik replied:
thanks for seeing that!

junniest commented on phylo/src/tkf_model/reestimate/tests.rs:
```diff
@@ -0,0 +1,103 @@
+use crate::{record_wo_desc as record, tree, Error};
+
+#[test]
+fn tkf_reestimate_with_out_choice() {
```
nitpick: it should probably be `without` instead of `with_out`

junniest commented on phylo/src/tkf_model/tkf92_additional_blocks.rs:
```diff
@@ -0,0 +1,261 @@
+            TKF92Parameters::Lambda => (f64::EPSILON, self.mu() - f64::EPSILON),
+            TKF92Parameters::Mu => (self.lambda() + f64::EPSILON, f64::MAX),
+            TKF92Parameters::R => PARAM_RANGE_UNIT_INTERVAL_EXCLUSIVE,
+            _ => panic!("Invalid parameter index for TKF model: {param:?}"),
```
This should be `bail!`

MattesMrzik replied:
Then we must also change the trait `ModelSearchCost` and `TKFModel.` Similarly, for the fn `set_param` of these traits it would make sense to return a Result. Should I go ahead and change it?

junniest replied:
Ah, of course, you're right. I think this is beyond the scope here for sure, but maybe we should change the trait to have `set_param() -> Result<()>` and `set_param_unchecked()` instead?

MattesMrzik replied:
See https://github.com/acg-team/rust-phylo/issues/155

junniest reviewed (approved):
Good job!

MattesMrzik commented:
> > Closes: #115, including #122. Still need to do #123
> 
> Is the description still accurate?

updated the description

MattesMrzik committed:
implemented some requested changes

github-actions[bot] committed:
Format Rust code using rustfmt

MattesMrzik committed:
point to github url to report bug in .expect

MattesMrzik committed:
fixed bug in doctest
